File: scripts/gen_combos.py
from collections import defaultdict

def guess_word(guess, answer):
    d = defaultdict(lambda: 0)
    ret = list("00000")

    for ch in answer:
        d[ch] += 1

    for i, ch in enumerate(answer):
        if guess[i] == answer[i]:
            ret[i] = "2"
            d[ch] -= 1

    for i, ch in enumerate(guess):
        if ret[i] == "2": continue

        if d[ch] != 0:
            ret[i] = "1"
            d[ch] -= 1
        else:
            ret[i] = "0"

    return ''.join(ret)

with open("all_words") as word_f:
    words_all = [x.strip() for x in word_f.readlines()]
with open("guess_words") as possible_f:
    words_possible = [x.strip() for x in possible_f.readlines()]

d = defaultdict(lambda: {})
total = len(words_possible)
for i, answer in enumerate(words_possible):
    if (not i%100):print(i, total)
    for guess in words_all:
        result = guess_word(guess, answer)
        d[guess][answer] = result

with open("guess_table", "w") as f:
    f.write("     ,")
    f.write(",".join(words_possible))
    f.write("\n")
    for guess in d:
        f.write(guess + ",")
        d_ = d[guess]
        f.write(",".join([d_[answer] for answer in words_possible]))
        f.write("\n")

    
    

File: scripts/gen_combos.cpp
#include <vector>
#include <string>
#include <fstream>
#include <sstream>
#include <iostream>
#include <array>

using namespace std;

vector<string> get_lines(const string &filename) {
    ifstream f(filename);
    string line;

    vector<string> ret;

    while (getline(f, line)) {
        ret.push_back(line);
    }

    return ret;
}

template <typename T, std::size_t N>
std::ostream &operator<<(std::ostream &os, const std::array<T, N> &container) {
    os << '[';
    for (int i = 0; i < N; ++i) {
        if (i != 0)
            os << ", ";
        os << container[i];
    }
    os << ']';
    return os;
}

array<int, 5> guess_word(const string &guess, const string &answer) {
    array<int, 5> ret = {0};
    array<int, 26> d = {0};

    for (char c : answer) {
        ++d[c - 'a'];
    }

    for (size_t i = 0; i < answer.size(); ++i) {
        char c = answer[i];
        if (guess[i] != c) continue;
        
        ret[i] = 2;
        --d[c - 'a'];
    }

    for (size_t i = 0; i < guess.size(); ++i) {
        char c = guess[i];

        if (ret[i] == 2) 
            continue;
        
        ret[i] = 0 <= --d[c - 'a'];
    }

    return ret;
}

int main() {
    vector<string> words_possible = get_lines("guess_words");
    vector<string> all_words=  get_lines("all_words");

    ofstream f("guess_table");
    f << "     ";
    for (const string &w : words_possible) {
        f << "," << w;
    }
    f << "\n";
    int cnt = 0;
    for (const string &g : all_words) {
        if (cnt % 100 == 0) {
            cout << cnt << " " << all_words.size() << "\n";
        }
        f << g;
        for (const string &w : words_possible) {
            f << ",";
            array<int, 5> res = guess_word(g, w);
            for (int i : res)
                f << i; 
        }
        f << "\n";
        ++cnt;
    }
}